File: 0051-n-queens/0051-n-queens.py
class Solution(object):
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        board=["."*n for i in range(n)]
        # board=[None]*n
        # s="."*n
        # for i in range(n):
        #     board[i]=s
        # Following are the only directions to check whether the queen exist or no
        upperDiagonal=[0]*(2*n - 1) # 2n-1 from the the max value for the formula n-1 + (col-row)
        lowerDiagonal=[0]*(2*n - 1) # 2n-1 from the max value for the formula col+row
        left=[0]*n
        res=[]
        
        def solve(col, upper,lower,lrow,res):
            if col==n:
                res.append(board[:])
                return 
            for i in range(n):
                if lrow[i]==0 and upper[n-1+col-i]==0 and lower[col+i]==0:
                    lrow[i]=1
                    upper[n-1+col-i]=1
                    lower[col+i]=1
                    board[i]=board[i][:col]+'Q'+board[i][col+1:]
                    solve(col+1,upper,lower,lrow,res)
        # Backtracking 
                    lrow[i]=0
                    upper[n-1+col-i]=0
                    lower[col+i]=0
                    board[i]=board[i][:col]+'.'+board[i][col+1:]
        
        solve(0,upperDiagonal,lowerDiagonal,left,res)
        return res

File: 0051-n-queens/0051-n-queens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        char [][] board = new char[n][n];
        
        for (int i = 0; i < n; i++){
            for (int j = 0; j < n; j++){
                board[i][j] = '.';
            }
        }
        List<List<String>>result = new ArrayList<>();
        int[] upperDiagonal = new int[(2 * n) - 1];
        int[] lowerDiagonal = new int[(2 * n) - 1];
        int[] leftRow = new int[n];
        
        helper(0, board, upperDiagonal, lowerDiagonal, leftRow, result, n);
        return result;
    }
    private List<String> construct(char[][] board, int n){
        List<String> res = new LinkedList < String>();
        
        for (int i = 0; i < n; i++){
            String s = new String(board[i]);
            res.add(s);
        }
        return res;
    }
    
    private void helper(int col, char[][] board, int[] upperDiagonal, int[] lowerDiagonal, int[] leftRow, List<List<String>> result, int n){
        // Base Case
        if (col == n){
            result.add(construct(board, n));
            return;
        }
        
        for (int i = 0; i < n; i++){
            if ((leftRow[i] != 1) && (upperDiagonal[n - 1 + col - i] != 1) && (lowerDiagonal[col + i] != 1)){
                
                // Recording the path
                board[i][col] = 'Q';
                leftRow[i] = 1;
                upperDiagonal[n - 1 + col - i] = 1;
                lowerDiagonal[col + i] = 1;
                helper(col + 1, board, upperDiagonal, lowerDiagonal, leftRow, result, n);
                
                // Backtracking
                board[i][col] = '.';
                leftRow[i] = 0;
                upperDiagonal[n - 1 + col - i] = 0;
                lowerDiagonal[col + i] = 0;

            }
        }
    }
}